File: modules/itac/jvm/src/main/java-igmore/edu/gemini/tac/qengine/qengine/skycalc/ElevationDecBinCalc.java
/*
 * Copyright (c) 2016-2025 Association of Universities for Research in Astronomy, Inc. (AURA)
 * For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
 */

//
// $
//

package edu.gemini.qengine.skycalc;

import edu.gemini.skycalc.Angle;
import edu.gemini.skycalc.Night;

import edu.gemini.skycalc.ElevationConstraintSolver;
import edu.gemini.skycalc.Interval;
import edu.gemini.skycalc.Solver;
import edu.gemini.skycalc.Union;
import edu.gemini.spModel.core.Site;
import jsky.coords.WorldCoords;

import java.util.*;

/**
 * Calculates dec bin percentages based upon airmass elevation constraints.
 */
public class ElevationDecBinCalc implements DecBinCalc {
    private final ElevationConfig conf;

    /**
     * Calculates DecBin percentages based upon the default configuration values.
     */
    public ElevationDecBinCalc() {
        this(ElevationConfig.DEFAULT);
    }

    /**
     * @param conf definition of when to consider a target observable
     */
    public ElevationDecBinCalc(ElevationConfig conf) {
        this.conf = conf;
    }

    public ElevationConfig getConfig() { return conf; }

    private List<Solver> getSolvers(Site site, List<WorldCoords> targets) {
        List<Solver> res = new ArrayList<Solver>();

        // targets.map(t => ElevationConstraintSolver.forAirmass(site, t, minAirmass, maxAirmass))
        for (WorldCoords wc : targets) {
            res.add(ElevationConstraintSolver.forAirmass(site, wc, conf.getMinAirmass(), conf.getMaxAirmass()));
        }
        return res;
    }

    // Gets the index of the bin that corresponds to the site.  All percentage
    // calculations are relative to this bin.
    private static int getSiteIndex(edu.gemini.spModel.core.Site site, DecBinSize size) {
        return ((int) Math.floor(site.latitude + 90.0)) / size.getSize();
    }

    @Override
    public List<Percent> calc(Site site, Date start, Date end, DecBinSize size, Angle ra) {
        List<WorldCoords> targets = size.genTargets(ra);
        List<Solver>      solvers = getSolvers(site, targets);

        long[] totals = new long[size.getBinCount()];

        Iterator<Night> it = new NightIterator(site, start, end, conf.getBounds());
        while (it.hasNext()) {
            int bin = 0;
            Night night = it.next();
            long startTime = night.getStartTime();
            long endTime   = night.getEndTime();
            for (Solver s : solvers) {
                Union<Interval> u = s.solve(startTime, endTime);
                for (Interval i : u.getIntervals()) {
                    totals[bin] += i.getLength();
                }
                ++bin;
            }
        }

        // Take as the max the value that corresponds to the dec overhead for
        // the site.
        long max = totals[getSiteIndex(site, size)];
//        for (long cur : totals) max = (cur > max) ? cur : max;

        List<Percent> res = new ArrayList<Percent>(size.getBinCount());
        for (long cur : totals) {
            res.add(new Percent(100.0 * ((double) cur / max)));
        }
        return res;
    }
}

File: modules/itac/jvm/src/main/scala/skycalc/ElevationDecBinCalc.scala
// Copyright (c) 2016-2025 Association of Universities for Research in Astronomy, Inc. (AURA)
// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause

package edu.gemini.qengine.skycalc;

import lucuma.core.enums.Site
import lucuma.core.math.Coordinates
import lucuma.core.math.skycalc.solver.Samples
import lucuma.core.math.BoundedInterval
import spire.math.extras.interval.IntervalSeq
import java.time.Instant
import lucuma.core.math.skycalc.solver.AirmassSolver
import lucuma.core.math.skycalc.ImprovedSkyCalc
import lucuma.core.math.RightAscension
import edu.gemini.tac.qengine.util.Percent
import spire.math.interval.Closed
import spire.math.interval.Open
import java.time.ZonedDateTime

/**
 * Calculates dec bin percentages based upon airmass elevation constraints.
 */
class ElevationDecBinCalc(conf: ElevationConfig) extends DecBinCalc {

  def getSolvers(site: Site, coords: List[Coordinates]): List[BoundedInterval[Instant] => IntervalSeq[Instant]] =
    coords.map: cs => 
      bi => 
        val samples = 
          Samples.atFixedRate(bi, java.time.Duration.ofMinutes(1)): i =>
            ImprovedSkyCalc(site.place).calculate(cs, i, false)
        AirmassSolver(conf.minAirmass, conf.maxAirmass).solve(samples)(bi)

  def calc(site: Site, start: ZonedDateTime, end: ZonedDateTime, size: DecBinSize, ra: RightAscension): List[Percent] =
    calc(site, start.toInstant(), end.toInstant(), size, ra)

  def calc(site: Site, start: Instant, end: Instant, size: DecBinSize, ra: RightAscension): List[Percent] = {
    val coords  = size.genCoordinates(ra)
    val solvers = getSolvers(site, coords)
    val totals  = Array.ofDim[Long](size.binCount)
    val it      = NightIterator.bounded(conf.bounds, site, start, end)
    while (it.hasNext) {
      var bin = 0
      val night = it.next
      solvers.foreach { s =>
        val u = s.apply(night.interval)
        u.intervalIterator.foreach: i =>
          
          val end = i.upperBound match
            case Closed(a) => a
            case Open(a) => a
            case b => sys.error(s"can't handle $b")
          
          val start = i.lowerBound match
            case Closed(a) => a
            case Open(a) => a
            case b => sys.error(s"can't handle $b")

          totals(bin) += end.toEpochMilli() - start.toEpochMilli()
          
          bin += 1
      }
    }    
    val max = totals(ElevationDecBinCalc.getSiteIndex(site, size))
    totals.toList.map: cur =>
      Percent(100.0 * cur.toDouble / max)
  }

}

object ElevationDecBinCalc:

    // Gets the index of the bin that corresponds to the site.  All percentage
    // calculations are relative to this bin.
    private def getSiteIndex(site: Site, size: DecBinSize): Int =
      math.floor(site.latitude.toAngle.toSignedDoubleDegrees + 90.0).toInt/ size.degrees
